Use "DocumentId" in indexes

jtkech opened the pull request:
Fixes #6966


Related to #7859 and #7757 where we discussed about unclustered indexes, here some parts of the discussion

> 1. As i saw, if we have an unclustered index on A1, A2, An, the only way to get is used is when all the required columns for the produced SELECT OR for a produced JOIN are included in A1, A2, An

> 2. Then, and **only if the above condition is true**, in the WHERE / OR / AND parts, **an unclustered index may be used if it includes some of the colums that are checked, and / or a clustered one for the other columns**

> 3. So, when we do a QueryIndex(...) as the produced select is on all columns, unclustered indexes are never used and seem to be useless, unless we define an index including all columns

> 4. Then, when we do a query on Document with an inner join on LocalizedContentItemIndex, so on DocumentId, **an unclustered index may be used if it includes DocumentId** (Note: then see 2. above)

> 5. **Yes you're right, but as i saw, this unclustered index including DocumentId is used when an index record is deleted**, not when inserted with new values and then updated with the right DocumentId. But yes, still useful on updating as we also need to delete previous record(s).

> 6. One rule could be to include DocumentId, then all columns but excluding those of too large size. Hmm, unless most of the colums (e.g. all columns) are not used in a joined query, as for AutoroutePartIndex where it's better to just include DocumentId, **knowing that when deleting an index record, only WHERE DocumentId is used**.

So it's related to the `CreateIndex()` we do for index tables in migrations, where `DocumentId` is important for joined query from `ContentItem` to `SomeIndexTable`, and also when deleting content items. Without at least `DocumentId`, unclustered indexes are not used, and it falls back to a clustered scan

    SchemaBuilder.AlterIndexTable<LocalizedContentItemIndex>(table => table
        .CreateIndex("IDX_LocalizationPartIndex_DocumentId",
            "DocumentId", // <= At least this one
            "LocalizationSet",
            "Culture",
            "ContentItemId",
            "Published",
            "Latest")
    );

jtkech commented:
@deanmarcussen I started working on this

- The rule being to only have one index including `DocumentId` and all other columns except those of a big size

    I saw that we limited some column sizes for mySql, e.g. to 768

    **But is it a limit per column or for the whole index?** If so we will still need to split some of them

- For migrations the rule being that `Create()` **include all alterations** so that it can shortcut all next steps

    For `ContentItemIndex` the migrations `Create()` was not shortcutting all the next steps, idem for the content fields 
migrations, here because `Create()` was not including the new `MultiTextFieldIndex`.

deanmarcussen commented:
> @deanmarcussen I started working on this
> 
> * The rule being to only have one index including `DocumentId` and all other columns except those of a big size
>   I saw that we limited some column sizes for mySql, e.g. to 768
>   **But is it a limit per column or for the whole index?** If so we will still need to split some of them

The limit is for the total of all columns in the entire index. MySql is the lowest 768. MySql will throw during index creation.
See these tests in YesSql 
https://github.com/sebastienros/yessql/blob/4fa577395240d1cfee59068bdb97e1b02ff17564/test/YesSql.Tests/MySqlTests.cs#L92-L275

and for SQL Server (which is higher 1700, and throws exceptions differently it gives a warning when creating, and throws when saving a large index.
https://github.com/sebastienros/yessql/blob/4fa577395240d1cfee59068bdb97e1b02ff17564/test/YesSql.Tests/SqlServerTests.cs#L152-L446

deanmarcussen commented on src/OrchardCore.Modules/OrchardCore.Alias/Migrations.cs:
```diff
@@ -33,11 +32,11 @@ public int Create()
 
             SchemaBuilder.AlterIndexTable<AliasPartIndex>(table => table
-                .CreateIndex("IDX_AliasPartIndex_Alias", "Alias", "Published", "Latest")
+                .CreateIndex("IDX_AliasPartIndex_DocumentId", "DocumentId", "Alias", "ContentItemId", "Latest", "Published")
```
For infos @jtkech this goes over 768. (which is why I never put the `ContentItemId` in the index originally.

I ran the functional tests against this branch. Here's the link https://github.com/OrchardCMS/OrchardCore/actions/runs/468293972 to the result. (do you know how to run them manually? Just go to actions, pick the functional tests, and choose your branch).

They're not perfect, as they only use a recipe, so things like `OpenId` don't get started, but still useful.

Locally, I'm just using a docker of mysql and postgres, and tweak some of the integration tests to initialize features, when necesary. We still need to get a better way of testing features. But extending the functional tests, to enable them, seems to be the preferred way.

jtkech commented:
@deanmarcussen 

> The limit is for the total of all columns in the entire index. MySql is the lowest 768

So we need to split some of them, but to do this i need to see how they are used, which was not the case when you include all columns. It would have been too simple ;)

jtkech commented:
@deanmarcussen

Okay, i did some mysql functional tests, here what i deduced about indexes sizes

- Here some data sizes (at least with the mysql version used in functional tests)

      Char     => 4 bytes
      Int      => 4 bytes
      Bool     => 1 byte

      // Depends on the fractional second precision
      DateTime => 5 + (0 up to 3) bytes => 5 bytes in our case but up to 8 bytes

- The maximum size of the sum of all columns is 3072 bytes, for example

      768 chars
      765 chars + 2 ints + 4 bools
      763 chars + 2 DateTime (if 5 bytes) + 2 ints + 2 bools

    Note: When an index has an unique column, i was able to define it up to 4000 chars * 4 = 16000 bytes

I will upate the indexes according to these results

deanmarcussen commented:
> But one problem is that we can't tweak the columns sizes of existing tables (not index tables), that would allow to create less index tables. So it becomes more tied to the current usage to find the best compromise.

Math all adds up to me @jtkech 

Not sure why you are trying to compress them all into one index though, i.e. `MultiTextField`. It is a compromise but I just went for what I thought was the best compromise there, i.e. three indexes, rather than compressing all of it into one.

(not fully thought through with the multitext, or tested hard, was trying to get it finished for @agriffard ). So there we could reduce sizes if useful (it's very recently in, so not an issue to change it a little bit)

But in short, yes, indexes are full of compromises.

Ok with reducing length of `Alias`. Hopefully doesn't break anyone.

jtkech commented:
@deanmarcussen 

> Not sure why you are trying to compress them all into one index though, i.e. MultiTextField. It is a compromise but I just went for what I thought was the best compromise there, i.e. three indexes, rather than compressing all of it into one.

So, because we always do a join on `DocumentId`, we include it in any index, otherwise it would be useless, so if we have multiple indexes for an index table, any of them can be selected and may not contain the columns related to the query

> (not fully thought through with the multitext, or tested hard, was trying to get it finished for @agriffard )

No problem, my comments are always just for info, never a reproach ;)

> So there we could reduce sizes if useful (it's very recently in, so not an issue to change it a little bit)
> Ok with reducing length of Alias. Hopefully doesn't break anyone.

Okay cool

---

**So, what we do with existing columns whose sizes are already 768 chars?** E.g. the `Text` column of the `TextFieldIndex`, we have 2 options for these columns

- Change the size even it is a breaking change, but i don't think so

- Don't define an index as it would be useless if not including `DocumentId`

Same kind of remark for some columns with a lower size but that still prevents us to use n of them

- E.g. we can't define an index including 4 columns of 255 chars, here i think we just need to exclude one of them, that's okay if one of them is not used, otherwise exclude the one that is less used.

    And without creating an additional index also including `DocumentId` because of the reason i mentioned the first part of my comment.

jtkech commented:
@deanmarcussen i did additional tests using mysql functional tests

So here, as a summary, some more info, and what i will do and suggest

- I confirm that a bool takes one byte, not physically but how it is taken into account for checking the max size.

- I will not change the size of any existing columns, unless for alias, and maybe for the multi text value.

- Try to define only one index per index table, if not possible try to only add one more based on the usage of queries, knowing that, in a joined query on `DocumentId`, if a selected index doesn't contain all columns that are used, even if another non clustered index contains the remaining ones, it will not be used.

- For any column that is alreay of 768 chars, i will not define any index that would not include `DocumentId`

deanmarcussen commented:
@jtkech sounds good.

Maybe yes for the multitext, it would be useful to have published and latest in there.

For the other field indexes, I imagined, that really for querying a value, the query might want to be more like this
`var query = Query.With<TextFieldIndex(x => x.Value == "foo").With<ContentItemIndex>(x => x.Published);` 

where a join is maybe better than no index. perhaps. 
I'm not sure how useful having `Published` and the other values in the index really where. However they are there now ;)

jtkech commented:
@deanmarcussen 

- I was not sure to change the alias size as it is there since one year, yes multi text is more recent

    There is not only alias where existing column sizes is a concern. E.g. while adding indexes to workflow, i saw that the Ids whose size is 26 chars use the default size of 255 chars, no luck ;)

- But there is another strategy, we could change some columns sizes to be able to add some indexes, but as rarely as possible, and only in the `Create()` method on a new schema, not in the related migration step that will not take the benefit of a new index, but this allows to take a better benefit on a new schema.

    In that case it is important that all migrations follow the pattern so that `Create()` is updated to do all the alterations and then always skip the following migration steps, which was not the case for some of them, e.g. the multi text.

> For the other field indexes, I imagined, that really for querying a value, the query might want to be more like this
var query = Query.With<TextFieldIndex(x => x.Value == "foo").With<ContentItemIndex>(x => x.Published);
> where a join is maybe better than no index. perhaps.

Good idea! And we could suggest this if a given index still doesn't include e.g. `Published`

> I'm not sure how useful having Published and the other values in the index really where. However they are there now ;)

For relatively low size indexes it may be better to have only one with all values than splitting them, but yes for other ones i will try to reduce their sizes in a second pass, but i think it's okay to always have the `Published` and `Latest` BITs.

E.g. for an existing column of 768 chars, if somewhere we change its size to include `DocumentId`, i think that's okay to reduce it's size also to include the `Published` and `Latest` BITs

Skrypt commented:
@jtkech Hmm, don't leave it there when you are done because this module is not so great.

jtkech commented:
@Skrypt Ah, i didn't know that

Anyway its just for testing its migrations by running the functional tests for mysql through github actions

jtkech commented:
@deanmarcussen 

I updated all migations and did some mysql func tests

- For infos, the default size of a string is 255 chars on the mysql we are using

 - For testing i added to a recipe e.g. the content fields sql queries, the mysql func tests still pass.

- We now can check if the size of some indexes can be reduced, if you think that normally some columns of a given set will not be used in a query, i will check it on my side this night.

jtkech commented:
@deanmarcussen i didn't do so much changes, see my previous comment that i updated (nothing urgent for me)

jtkech commented:
@deanmarcussen @Skrypt @sebastienros for infos

I think i saw all we can do to finalize this PR, it only depends on some options that we have to choose. The main concern is related to the column sizes of some index tables, but only for some of them, that's okay for the other ones.

- We only talk about queries generated by YesSql starting with `SELECT DISTINCT ... FROM Document` (without `distinct` the results are not the same), then `INNER JOIN IndexTable ON DocumentId ...`, then `WHERE ...`.

    - To be selected a non-clustered index need at least to include `DocumentId`, otherwise it is useless.

    - Only one non-clustered index, so including `DocumentId`, can be selected for a given inner join

    -  If the selcted index includes all columns used in the `WHERE ...`, no scan on the clustered index is done.

    -  Otherwise a scan is done on the clustered index, even if the missing columns are in other indexes.

    -  But the columns of the selected index are still used, so at this point that's okay to find good compromises.

- BUT, like with the inner join, if the query has an `ORDER BY`, each column in it needs to be included in the non-clustered index, otherwise it is not used at all, even if it contains columns used in the previous `WHERE ...`.

    - We could say to people to sort on their side, but better to not do this when the query includes a `TOP`.

    - So, in place of having too many indexes case by case, one solution is to also have some indexes with a little more columns, but here a compromise is harder to find with the current column sizes and the index size limit.

    - Always good to reduce the index sizes resulting in less sql pages to manage them, but see next section.

        - E.g. we have some identifiers of 26 chars that use the default of 255 chars => up to 1020 bytes.
        - So, just using 3 such columns => 3060 bytes which is already closed to the max index size for mysql.
        - E.g. `ContentType`, `Owner`, `Author` also use 255 chars that may be reduced.

- To reduce the size of a column, we first need to drop the non-clustered indexes that uses this column, but it doesn't work on some sql server versions. In fact, for sql server, we generate an `IF EXISTS` which is not always avalaible.

      IF EXISTS
      Applies to: SQL Server ( SQL Server 2016 (13.x) through current version).

    - Note: Maybe on YesSql side we could use `SupportsIfExistsBeforeTableName => false`.

    - Note: Dropping an index works with sqlite, didn't try with other database providers.

- So, we could have the following options (but see next updated section)

    - Mainly cover queries without an `ORDER BY`, and without reducing the index sizes.

    - Also cover `ORDER BY`, without reducing the index sizes, but maybe too many / large indexes.

    - Also reduce the index sizes by reducing the column sizes of the related index table.

        - Allowing new lighter indexes or with more columns, but only in the migration `Create()`.

        - Also in a migration step, but without supporting e.g. SQL server before 2016 (13.x).

- **Update**

    - Altering a column doesn't work with SQLite, so we can't reduce column sizes in migration seps.

    - I would opt to still reduce some column sizes to optimize indexes, but only on `Create()` a new schema.

    - Then on existing databases, add incomplete indexes but still better than the previous ones that are useless.

    - **Note: The above is only related to index tables where the column sizes is a concern.**

    - Finally, about dropping useless indexes, can be done by people manually

    - Last option, as @Skrypt suggested, more advanced migration scripts, maybe per provider. Not in this PR as it seems to be quite complex, but something that we could plan to do.

deanmarcussen reviewed (commented):
@jtkech looking good, comprises always, but you have made some good ones I think.

Couple of comments, I think maybe users is not quite right?

deanmarcussen commented:
For infos https://github.com/sebastienros/yessql/issues/322

I took a review of one of our running prod Azure Orchard Core installations, and so far (it's only been up a couple of weeks), it hasn't collected any other recommendations.

We probably still need to do the content item index table at some point, but another pr.

jtkech commented on src/OrchardCore.Modules/OrchardCore.Users/Migrations.cs:
```diff
@@ -27,86 +27,90 @@ public int Create()
                 null);
 
-            // Return 8 here to skip migrations on new database schemas.
-            return 8;
```
We did this because in the above `Create()`, we were using `return UpdateFrom1()`, which is not a good pattern.

After adding a migration step, `Create()` should be updated to do all the things, and always shortcut migration steps.

This is what i did above by adding in `Create()` the columns added in `UpdateFrom2()`, and then shortcut the steps.

deanmarcussen replied:
Ah I see, yes it was a bad / weird pattern, so you have removed it. great.

jtkech commented on src/OrchardCore.Modules/OrchardCore.ContentFields/Indexing/SQL/MultiTextFieldIndexProvider.cs:
```diff
@@ -12,8 +12,10 @@ namespace OrchardCore.ContentFields.Indexing.SQL
+        // Maximum length that MySql can support in an index under utf8 collation is 768,
+        // minus 1 for the `DocumentId` integer (character size = integer size = 4 bytes).
+        // minus 1 (freeing 4 bytes) for the additional 'Published' and 'Latest' booleans.
+        public const int MaxValueSize = 766; // 
```
Done on my side ;)

jtkech commented:
> We probably still need to do the content item index table at some point, but another pr.

The indexes of the `ContentItemIndex` table are already updated in this PR, but maybe you meant something else

deanmarcussen commented:
> The indexes of the `ContentItemIndex` table are already updated in this PR, but maybe you meant something else

I just meant we still need to look at getting the `ContentType` in there somewhere. https://github.com/OrchardCMS/OrchardCore/issues/6966

But happy for that to be another pr.

You still have this as not ready? For me it looks good to merge.

jtkech commented:
@deanmarcussen 

> I just meant we still need to look at getting the ContentType in there somewhere. #6966

Already done here through 3 indexes (maybe too much, see below), e.g. with this index
https://github.com/OrchardCMS/OrchardCore/blob/b4ea2b5818aca47e73ed580d8c033bbb291385ad/src/OrchardCore/OrchardCore.ContentManagement/Records/Migrations.cs#L74

> You still have this as not ready? For me it looks good to merge.

Yes, it is quite ready, just need some last tweaks, but since one week i always do and revert my changes ;)

The last thing i'm thinking of is, for perf reasons, to reduce the number /sizes of some indexes

As a summary about a given non clustured index as we are using it

- Need to include `DocumentId` because of the `INNER JOIN`, **that's okay**.

- Need to include all or part of the columns (at least one) used in the `WHERE`, so that it is selected and used in the query, even it doesn't include all columns, so we can find good compromises, **here also that's okay**.

- The main point is when using `ORDER BY`, to be selected an index need to not only to include the columns of the `ORDER BY`, but also of the `WHERE`, **so it needs to be a fully covered index, without any compromise**.

    So, just because of this we need more and wider indexes including the same subset of columns.

    So i'm thinking about not covering all possible order by, e.g. Look at the 4 indexes i defined on `ContentItemIndex`
https://github.com/OrchardCMS/OrchardCore/blob/b4ea2b5818aca47e73ed580d8c033bbb291385ad/src/OrchardCore/OrchardCore.ContentManagement/Records/Migrations.cs#L33

     I'm thinking first about not covering `Owner` and `Author`, more intended to be used in a `WHERE`, and where the optimizer will use a compromise (see above). Then, not covering order by `DisplayText`, but still by `CreatedUtc` and other UTC dates. This would allow to just define the 2 following indexes, in place of 4 wider indexes.

            SchemaBuilder.AlterIndexTable<ContentItemIndex>(table => table
                .CreateIndex("IDX_ContentItemIndex_DocumentId",
                    "DocumentId",
                    "ContentItemId",
                    "ContentItemVersionId",
                    "Latest",
                    "Published")
            );

            SchemaBuilder.AlterIndexTable<ContentItemIndex>(table => table
                .CreateIndex("IDX_ContentItemIndex_DocumentId_ContentType",
                    "DocumentId",
                    "ContentType",
                    "CreatedUtc",
                    "ModifiedUtc",
                    "PublishedUtc",
                    "Latest",
                    "Published")
            );

deanmarcussen commented:
Ok, cool, sorry I missed the addition of `ContentType`.

I would think about it from the concept of what we are already providing in the admin controller for contents.
in order of preference
- ContentType (is on every search, because it is an `IsIn`)
- Published/Modified (is on every search)
- DisplayText for filtering, important, probably less so for ordering?

So maybe as display text is less common, to be less indexed is ok.

But I would still probably go for the four indexes. 
The idea being to optimize reads over writes.

So adding the more indexes will make writes more expensive, but **should** improve reads, which as a general concept in the CMS we do way more reads than we do writes.

I would drop the one including `Author`. We don't ever search for Author, it's really display only, unless you can think of a need?

Bearing in mind, users can still optimize and create their own indexes where it is useful for their queries.

jtkech commented:
@deanmarcussen Thanks, useful infos

> The idea being to optimize reads over writes.

Yes, but depending on the usage of queries, there are other rules that may be good also for read perf , e.g. less and lighter indexes, easier to select and use, don't index small tables, seeking doesn't always mean it is better than scanning.

> I would drop the one including Author

So, if i understand correctly, you agreed to remove the one including DisplayText, right?
Then you are suggesting to also remove the one including Author.
**What's your feeling about the one including Owner?**

> Bearing in mind, users can still optimize and create their own indexes where it is useful for their queries.

Or delete existing ones ;)

So is it better to let them delete some if we added too many, or let them adding missing ones.

I'll work on it this night, maybe i will just let it as it is ;)

jtkech commented:
Didn't have so much time but i did some tests

For now, with one hundred of items i can see the benefit of covering indexes, with one thousand it is less significant, even we have less io reads on the index, but still the same number of reads on the main document (always the higher number),  but the total amount of reads is less,  so it may depend on the storage system.

Then, when using an order by, using a covering index or not seems to be less significant

But i didn't have so much time, so i will do more tests and finalize this PR this week end

jtkech commented:
@Skrypt @deanmarcussen For infos

Just tried in my local branch to add the following in `TextFieldIndexProvider`, so that each time we publish a new version we don't keep the previous one that is soft deleted, this to prevent big index tables, e.g. if you have many items and / or that are frequently published, even more if you have multiple fields on some content type.

                    if (!contentItem.Published && !contentItem.Latest)
                    {
                        return null;
                    }

So, i think i will do this on all field index providers, and check if there are other ones where we could do the same.

Skrypt commented:
This was not how it was done in 01. We we're indexing every textfield versions. Maybe it should be an option of that module then.

jtkech commented:
@Skrypt Okay thanks, so i think that for now i will let it as it is ;)

jtkech commented:
So okay, for infos about the `WorkflowStatus` enum for PostgreSQL

The concern being to add, on the `WorkflowIndex` table, a new index including the `WorkflowStatus` column

To make it working we need to map the `WorkflowStatus` to an int, as already done in the dev branch

Note: We would not need this if the `WorkflowStatus` column was of int type when the table was created, but we can't alter a column type in a migration step.

deanmarcussen commented:
> To make it working we need to map the WorkflowStatus to an int, as already done in the dev branch

Sorry, confused, did you keep it as an int in this PR, as the data column is of type string, and it doesn't work to change it?

That's fine with me, it is what it is

jtkech commented:
> Sorry, confused, did you keep it as an int in this PR, as the data column is of type string, and it doesn't work to change it?

Yes

And i removed the not ready label

jtkech commented:
Just saw last little tweaks that i will do this night

deanmarcussen commented:
Did you finish your tweaks @jtkech ? 

Ready to merge?

jtkech commented:
Yes as i removed the not ready label

Let's go, will see how it behaves ;)
